servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+
+        // protected by emitting lock, or only accessed in side the Subscriber thread
+        private boolean targetTerminated;
+        private int upstreamDemand;
+        @Nullable
+        private Subscription subscription;
+
+        private final Subscriber<? super R> target;
+        private final Queue<Object> signals;
+        private final PublisherFlatMapMerge<T, R> source;
+        private final Set<FlatMapPublisherSubscriber<T, R>> subscribers;
+
+        FlatMapSubscriber(PublisherFlatMapMerge<T, R> source, Subscriber<? super R> target) {
+            this.source = source;
+            this.target = target;
+            // Start with a small capacity as maxConcurrency can be large.
+            signals = newUnboundedMpscQueue(min(2, source.maxConcurrency * source.maxMappedDemand));
+            subscribers = newSetFromMap(new ConcurrentHashMap<>());
+        }
+
+        @Override
+        public void cancel() {
+            doCancel(true);
+        }
+
+        @Override
+        public void request(final long n) {
+            assert subscription != null;
+            if (!isRequestNValid(n)) {
+                subscription.request(n);
+            } else if (pendingDemandUpdater.getAndAccumulate(this, n,
+                    FlowControlUtils::addWithOverflowProtectionIfNotNegative) == 0) {
+                drainPending();
+            }
+        }
+
+        @Override
+        public void onSubscribe(final Subscription s) {
+            if (!checkDuplicateSubscription(subscription, s)) {
+                return;
+            }
+
+            // We assume that FlatMapSubscriber#cancel() will never be called before this method, and therefore we
+            // don't have to worry about being cancelled before the onSubscribe method is called.
+            subscription = ConcurrentSubscription.wrap(s);
+            target.onSubscribe(this);
+
+            upstreamDemand = source.maxConcurrency;
+            subscription.request(upstreamDemand);
+        }
+
+        @Override
+        public void onNext(@Nullable final T t) {
+            final Publisher<? extends R> publisher = requireNonNull(source.mapper.apply(t));
+            FlatMapPublisherSubscriber<T, R> subscriber = new FlatMapPublisherSubscriber<>(this);
+            final boolean added = subscribers.add(subscriber);
+            assert added; // FlatMapPublisherSubscriber relies upon object equality and should never fail.
+            for (;;) {
+                final int prevActiveSources = activeMappedSources;
+                if (prevActiveSources < 0) {
+                    // We have been cancelled, or already completed and the active count flipped to negative, either way
+                    // we don't want to Subscribe or retain a reference to this Publisher.
+                    subscribers.remove(subscriber);
+                    break;
+                } else if (activeMappedSourcesUpdater.compareAndSet(this, prevActiveSources, prevActiveSources + 1)) {
```

Comment:
I wonder if we could just use `incrementAndGet(...)` and decrement again if its not what we expect. This should be cheaper as this will use XADD when on x86

Something like:

```java
            for (;;) {
                 final int prevActiveSources = activeMappedSources;
                 if (prevActiveSources < 0) {
                     // We have been cancelled, or already completed and the active count flipped to negative, either way
                     // we don't want to Subscribe or retain a reference to this Publisher.
                     subscribers.remove(subscriber);
                     break;
                 // try to optimistically increment and if the outcome is not what we expect rollback
                 } else  if (activeMappedSourcesUpdater.incrementAndGet(this) ==  prevActiveSources + 1) {
                     publisher.subscribeInternal(subscriber);
                     break;
                 } else {
                     // rollback
                     activeMappedSourcesUpdater.decrementAndGet(this)
                 }
             }

Reply:
negative values are meaningful as we use the negative values to signal completion to the mapped sources `terminateActiveMappedSources`, so optimistically incrementing may result in premature completion. lets investigate this optimization as a followup PR.

Reply:
actually we don't expect `onNext` after termination so we may be able to use this approach. However this code is using consistent approach with `PublisherFlatMapSingle` and we should keep them inline ... good candidate for followup PR!